Request render optin mode

ggetz opened the pull request:
Created opt-in `requestRenderMode` which will only render the scene when needed. Changes it watches for are currently minimal, but can be triggered as needed by calling `requestRender()`.

* Added events  to `TaskProcessor` on a successful completed web worker task and `RequestScheduler` on a successful request that the scene can watch for. (I'm thinking these might be better off private?)
* Camera changes and window resizing will cause a render
* If the simulation time change exceeds `scene.maximumRenderTimeChange`, defaults to `0.5` and can be set to `undefined` to never change based on simulation time.
* new public function `scene.requestRender` for forcing render in this mode

`debugShowFrames` per second might be misleading, the updates are still happening at a higher frame rate than what's being rendered, and the fps value is not updated until after a new render. Should I update the display to reflect this or leave it as is?

cesium-concierge commented:
![Signed CLA is on file.](https://img.shields.io/badge/CLA-signed-brightgreen.svg)

@ggetz, thanks for the pull request! Maintainers, we have a signed CLA from @ggetz, so you can review this at any time.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

pjcozzi commented:
> Added events to TaskProcessor on a successful completed web worker task and RequestScheduler on a successful request that the scene can watch for. (I'm thinking these might be better off private?)

In general, if we don't have a concrete use case for making something public, keep it private.  Less API surface area is less work and less maintenance.

pjcozzi commented:
> debugShowFrames per second might be misleading, the updates are still happening at a higher frame rate than what's being rendered, and the fps value is not updated until after a new render. Should I update the display to reflect this or leave it as is?

I don't 100% follow.  Do you mean that the fps might be basically 0 when nothing is happening or it, for example, 3 when the scene is only rendered once in awhile due to tile downloads, etc.?

If so, perhaps report:
* `n fps` when rendering full speed
* `n/a fps` when not rendering
* `n fps (throttled)` when rendering on demand

pjcozzi commented:
Part of #1865

pjcozzi commented:
@wallw-bits @chris-cooper @willemvdg @kring testing and code reviews are appreciated here, thanks!

pjcozzi commented on Apps/Sandcastle/gallery/Scene Rendering Performance.html:
```diff
@@ -0,0 +1,123 @@
+<!DOCTYPE html>
```
Is there any screenshot or graphic that could be used for this Sandcastle example?

pjcozzi commented on CHANGES.md:
```diff
@@ -3,6 +3,9 @@ Change Log
 
 ### 1.41 - 2017-01-02
 
+* Added optional scene request render mode to reduce CPU usage
```
Given the holiday, I don't think this will make the 1.41 release.

pjcozzi commented on Source/Scene/Scene.js:
```diff
@@ -3563,5 +3649,17 @@ define([
+     * @see Scene#requestRenderMode
+     */
+    Scene.prototype.requestRender = function() {
+        if (!this._isRendering) {
```
Why check, just set it to `true`.

pjcozzi commented on Source/Scene/Scene.js:
```diff
@@ -709,6 +719,37 @@ define([
         this._aspectRatioVR = undefined;
 
+        /**
+         * When <code>true</code>, rendering a frame will only occur when needed as determined by changes within the scene.
```
Please provide more detail here just like you did in the first comment in this pull request.  In particular, make sure that the reader understands that they will need to explicitly call `requestRender` under many cases.

pjcozzi commented on Source/Scene/Scene.js:
```diff
@@ -709,6 +719,37 @@ define([
+        this._isRendering = true;
+
+        /**
+         * If {@link requestRenderMode} is <code>true</code>, this value defines the maximum change in
```
Provide more context and implications - this will impact sun lighting, water animation, and the choppiness of CZML animation, for example.

pjcozzi commented on Source/Scene/Scene.js:
```diff
@@ -709,6 +719,37 @@ define([
+         * @default false
+         */
+        this.requestRenderMode = defaultValue(options.requestRenderMode, false);
+        this._isRendering = true;
```
Is `_isRendering` the best name?  Should this be `_requestRender`, `_renderDirty`, etc.?

ggetz commented:
Fixed test failures and addressed comments from @pjcozzi

lilleyse commented on Source/Scene/Scene.js:
```diff
@@ -709,6 +719,40 @@ define([
+        this._renderRequested = true;
+
+        /**
+         * If {@link requestRenderMode} is <code>true</code>, this value defines the maximum change in
```
This link doesn't work.

lilleyse commented on Source/Scene/Scene.js:
```diff
@@ -2829,44 +2905,51 @@ define([
+        updateDebugShowFramesPerSecond(this, shouldRender);
+        RequestScheduler.update();
+        callAfterRenderFunctions(this._frameState);
+        this._postRender.raiseEvent(this, time);
```
Should the after render functions and events only happen if a render has actually occurred? They should probably stay here, but I'm not sure.

ggetz replied:
If the listeners subscribed to these events do not take into account time in the callback function, this could cause some unexpected issues. If they do take into account time, it will advance correctly even when render was not called. To be safe, I would prefer to leave them here, but naming is not ideal. We could rename them from "render" to "update"?

lilleyse replied:
Ok, It might be worth updating the doc for `scene.postRender` to indicate that the event is raised every `update` instead of each `render`. I wonder if that qualifies as a breaking change...?

ggetz replied:
It won't break any functionality without explicitly enabling `requestRenderMode`. Should I just note that in the documentation then? I'm not sure either if this qualifies as a breaking change. @pjcozzi ?

pjcozzi replied:
Please review the use cases for this event and decide the right course of action - seems that it currently couples render and perhaps time change and/or update - which may call for separate events and reasonable deprecation.  No need for anymore input from me.

lilleyse commented on Source/Scene/Scene.js:
```diff
@@ -3538,6 +3621,13 @@ define([
+        }
+        if (defined(this._removeTaskProcessorListenerCallback)) {
+            this._removeTaskProcessorListenerCallback();
+        }
```
These should be defined always.

lilleyse commented on Source/Core/RequestScheduler.js:
```diff
@@ -154,6 +165,7 @@ define([
             --numberOfActiveRequestsByServer[request.serverKey];
             request.state = RequestState.RECEIVED;
             request.deferred.resolve(results);
+            requestLoadedEvent.raiseEvent();
```
There is one other area where the `raiseEvent` should be triggered.

```
        if (isDataUri(request.url) || isBlobUri(request.url)) {
            request.state = RequestState.RECEIVED;
            return request.requestFunction();
        }
```

This is a quick path for data uris and blob uris.

lilleyse replied:
Also add a test for this case.

lilleyse commented on Source/Scene/Scene.js:
```diff
@@ -3563,5 +3653,15 @@ define([
     };
 
+    /**
+     * Requests a new rendered frame when {@link requestRenderMode} is set to <code>true</code>.
```
Same note about the link not working.

lilleyse commented on Source/Scene/Scene.js:
```diff
@@ -3563,5 +3653,15 @@ define([
+     */
+    Scene.prototype.requestRender = function() {
+        this._renderRequested = true;
+    };
```
For organizational purposes this and `cartesianToCanvasCoordinates` should come before `isDestroyed`, and preferably closer to similar areas of the code.

lilleyse commented on Specs/Core/TaskProcessorSpec.js:
```diff
@@ -124,4 +124,43 @@ defineSuite([
+        };
+        var eventRaised = false;
+        var removeListenerCallback = TaskProcessor.taskCompletedEvent.addEventListener(function () {
+            console.log('2');
```
Remove this.

lilleyse commented on Source/Scene/Scene.js:
```diff
@@ -2829,44 +2905,51 @@ define([
+
+        var shouldRender = !this.requestRenderMode || this._renderRequested || cameraChanged || (this.mode === SceneMode.MORPHING);
+
+        var now = JulianDate.clone(time);
```
Use a scratch variable instead, and clone into `_lastRenderTime` inside `if (shouldRender)`.

ggetz commented:
Thanks @lilleyse! Updated according to your comments.

Also,
> if the camera doesn't move or things aren't loading the animations ticks every 0.5 seconds

The `maxiumRenderTimeChange` controls this update rate, so yo can do it less frequently (or not at all) if your scene has no animations, or more often if you need smoother animations.

ggetz commented on Apps/Sandcastle/gallery/Scene Rendering Performance.html:
```diff
@@ -0,0 +1,123 @@
+function startup(Cesium) {
+    'use strict';
+//Sandcastle_Begin
+var viewer = new Cesium.Viewer('cesiumContainer', {
```
I added a field with a timestamp of when the frame was last rendered.

lilleyse commented:
Cool I'll look at the recent updates.

> The maxiumRenderTimeChange controls this update rate, so yo can do it less frequently (or not at all) if your scene has no animations, or more often if you need smoother animations.

Yup, I was just stating that things were operating correctly.

lilleyse commented:
There are a coupe eslint errors.

ggetz commented:
@lilleyse Sorry, they're all cleaned up!

lilleyse commented on Source/Scene/Scene.js:
```diff
@@ -218,7 +218,10 @@ define([
+     * @param {Boolean} [options.requestRenderMode=false] If true, rendering a frame will only occur when needed as
+     *  determined by changes within the scene. Enabling improves performance of the application, but requires using
+     *  {@link Scene#requestRender} to render a new frame explicitly in this mode. This will be necessary in many cases
+     *  after making changes to the scene in other parts of the API.
```
I think we typically keep parameter doc on a single line even it's long.

lilleyse commented on Source/Scene/Scene.js:
```diff
@@ -2923,15 +2942,14 @@ define([
         try {
             if (shouldRender) {
-                this._lastRenderTime = now;
+                this._lastRenderTime = JulianDate.clone(time, scratchTime);;
```
Actually now that I see the updated changes just remove the scratch and call `JulianData.clone(time, this._lastRenderTime)`.

lilleyse commented:
It looks like a couple review comments from the latest review got hidden, make sure to expand all.

lilleyse commented:
I wonder if there's more we can do to improve globe/imagery loading. For example, a simple Sandcastle like:

```
var viewer = new Cesium.Viewer('cesiumContainer', {
    requestRenderMode : true,
    maximumRenderTimeChange : 100
});
```
will not show the globe. I'm guessing globe's `render` and `update` are tightly coupled, and am curious if there's anything we can do about that.

lilleyse commented:
Missed this on the first go, but `Viewer` and `CesiumWidget` should have tests, including a test for resize if possible,

ggetz commented:
> I wonder if there's more we can do to improve globe/imagery loading. For example, a simple Sandcastle like ... will not show the globe. I'm guessing globe's render and update are tightly coupled, and am curious if there's anything we can do about that.

This is exactly what my next PR addresses. The tile loading for the globe was tied to the render loop, so nothing was getting loaded without rendering a new frame. I separated the update/render logic there, so the globe always loads in completely without time changes triggering the render.

lilleyse commented:
> This is exactly what my next PR addresses. The tile loading for the globe was tied to the render loop, so nothing was getting loaded without rendering a new frame. I separated the update/render logic there, so the globe always loads in completely without time changes triggering the render.

Ok, awesome!

ggetz commented:
Addressed your additional feedback @lilleyse, only https://github.com/AnalyticalGraphicsInc/cesium/pull/6065#discussion_r159514539 is outstanding.

biomassives commented:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (100/100 (100%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from o10.sgmail.github.com ([167.89.101.201]:29897)
	by chi-server32.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.89_1)
	(envelope-from <bounces+848413-b37a-greg=biomassiv.es@sgmail.github.com>)
	id 1eWpkS-0034Vt-Nh
	for greg@biomassiv.es; Wed, 03 Jan 2018 14:34:52 -0600
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed/relaxed; d=github.com; 
	h=from:reply-to:to:cc:in-reply-to:references:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=s20150108; bh=mc22SvfW3D7MgabdU5SXjPwmPtc=; b=IfoaQaOTOKLAh8fl
	gg7UrOPHWNHCdqdSsVmubUH41qbzp8MkYI4qBS+5Ljqw0vZB8KGwn4uGsDUVTcls
	AztsY9fmrxoNMzYC1ls3dt/JZb7WI9h5IGRUL1PBxRKPSfSOMgehgiig7HWllBQ8
	7381Q4Z8d8BA/Brh6hL01BxvPDI=
Received: by filter0607p1mdw1.sendgrid.net with SMTP id filter0607p1mdw1-32046-5A4D3E45-2
        2018-01-03 20:34:13.042847585 +0000 UTC
Received: from github-smtp2a-ext-cp1-prd.iad.github.net (github-smtp2a-ext-cp1-prd.iad.github.net [192.30.253.16])
	by ismtpd0005p1iad1.sendgrid.net (SG) with ESMTP id XSfxx2oUTDaXPc3qrrfjKA
	for <greg@biomassiv.es>; Wed, 03 Jan 2018 20:34:13.034 +0000 (UTC)
Date: Wed, 03 Jan 2018 20:34:13 +0000 (UTC)
From: Gabby Getz <notifications@github.com>
Reply-To: AnalyticalGraphicsInc/cesium <reply@reply.github.com>
To: AnalyticalGraphicsInc/cesium <cesium@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Message-ID: <AnalyticalGraphicsInc/cesium/pull/6065/c355119583@github.com>
In-Reply-To: <AnalyticalGraphicsInc/cesium/pull/6065@github.com>
References: <AnalyticalGraphicsInc/cesium/pull/6065@github.com>
Subject: Re: [AnalyticalGraphicsInc/cesium] Request render optin mode (#6065)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_5a4d3e44e30e9_118e3fa044368f3422653d";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: ggetz
X-GitHub-Recipient: biomassives
X-GitHub-Reason: subscribed
List-ID: AnalyticalGraphicsInc/cesium
 <cesium.AnalyticalGraphicsInc.github.com>
List-Archive: https://github.com/AnalyticalGraphicsInc/cesium
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+0042d4e268185d2bc3878d13dac03f49f03916ae03082e8a92cf000000011665004492a169ce10e3793d@reply.github.com>,
 <https://github.com/notifications/unsubscribe/AELU4kBX2frw3xnjEEaPw3teeueLWsy3ks5tG-REgaJpZM4RHcZ8>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: greg@biomassiv.es
X-SG-EID: ZlaDthyue/4HG+EgTb7JoZYs48JNb2iQtCcPV6oIs9XYKhieKS3+iBCehGurptg8G12Ky1tJR8MXnz
 ZfKCLHSQvFtqJNCj0MB/4JJV0h5N8DIpF+yDXTCjczom+iQkJMFB8oF3MROOMlLufoFKp0G7CZCjfL
 y2Dlms/jZGyDYTaOPVVHkNKQtxLBVFu1lNFCsS/lZlX+lvcLtGrrYzSOuw==

----==_mimepart_5a4d3e44e30e9_118e3fa044368f3422653d
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

Addressed your additional feedback @lilleyse, only https://github.com/AnalyticalGraphicsInc/cesium/pull/6065#discussion_r159514539 is outstanding.

-- 
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/AnalyticalGraphicsInc/cesium/pull/6065#issuecomment-355119583
----==_mimepart_5a4d3e44e30e9_118e3fa044368f3422653d
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

<p>Addressed your additional feedback <a href="https://github.com/lilleyse" class="user-mention">@lilleyse</a>, only <a href="https://github.com/AnalyticalGraphicsInc/cesium/pull/6065#discussion_r159514539" class="issue-link js-issue-link" data-error-text="Failed to load issue title" data-id="283343165" data-permission-text="Issue title is private" data-url="https://github.com/AnalyticalGraphicsInc/cesium/issues/6065">#6065 (comment)</a> is outstanding.</p>

<p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">&mdash;<br />You are receiving this because you are subscribed to this thread.<br />Reply to this email directly, <a href="https://github.com/AnalyticalGraphicsInc/cesium/pull/6065#issuecomment-355119583">view it on GitHub</a>, or <a href="https://github.com/notifications/unsubscribe-auth/AELU4j3LXOjHTcCsX2UJDp1lQlqnTb1Rks5tG-REgaJpZM4RHcZ8">mute the thread</a>.<img alt="" height="1" src="https://github.com/notifications/beacon/AELU4nRzCFsMqjbkEADB5dK9nCLiHcR_ks5tG-REgaJpZM4RHcZ8.gif" width="1" /></p>
<div itemscope itemtype="http://schema.org/EmailMessage">
<div itemprop="action" itemscope itemtype="http://schema.org/ViewAction">
  <link itemprop="url" href="https://github.com/AnalyticalGraphicsInc/cesium/pull/6065#issuecomment-355119583"></link>
  <meta itemprop="name" content="View Pull Request"></meta>
</div>
<meta itemprop="description" content="View this Pull Request on GitHub"></meta>
</div>

<script type="application/json" data-scope="inboxmarkup">{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/AnalyticalGraphicsInc/cesium","title":"AnalyticalGraphicsInc/cesium","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/AnalyticalGraphicsInc/cesium"}},"updates":{"snippets":[{"icon":"PERSON","message":"@ggetz in #6065: Addressed your additional feedback @lilleyse, only https://github.com/AnalyticalGraphicsInc/cesium/pull/6065#discussion_r159514539 is outstanding."}],"action":{"name":"View Pull Request","url":"https://github.com/AnalyticalGraphicsInc/cesium/pull/6065#issuecomment-355119583"}}}</script>
----==_mimepart_5a4d3e44e30e9_118e3fa044368f3422653d--

ggetz commented:
@lilleyse Updated the events (and changed the structure of the render function to reflect that. The update function is empty, but that is where the majority of the refactored code will go for the globe updates talked about above.). I added `preUpdate` and `postUpdate` events, and check for rendering after these events are triggered. `preRender` and `postRender` are only triggered when the scene renders a new frame. Updated docs and `CHANGES.md` to reflect the changes.

lilleyse commented:
Ok, I will look over the new changes a bit later today.

lilleyse commented on CHANGES.md:
```diff
@@ -3,9 +3,12 @@ Change Log
-* Added optional scene request render mode to reduce CPU usage
-    * `scene.requestRenderMode` enables a mode which will only request new render frames on changes to the scene, or when the simulation time change exceeds `scene.maximumRenderTimeChange`.
+* Breaking changes
+    * `scene.preRender` and `scene.postRender` events are called immediately before and after scene rendering only if the scene renders a frame. See `scene.requestRenderMode`. Use `scene.preUpdate` and `scene.postUpdate` for task that require regular updates.
```
I actually don't think this needs to be a breaking change. Though the connection between render and update events still needs to be clear. Maybe below write:

> Added `scene.preUpdate` and `scene.postUpdate` events that are raised before and after the scene updates respectively. The scene is always updated before executing a potential render. Continue to listen to `scene.preRender` and `scene.postRender` for when the scene renders a frame.

lilleyse replied:
Capitalize `Scene` for all these.

lilleyse commented on Source/Scene/FrameRateMonitor.js:
```diff
@@ -98,7 +98,7 @@ define([
         this._pauseCount = 0;
 
         var that = this;
-        this._preRenderRemoveListener = this._scene.preRender.addEventListener(function(scene, time) {
```
There are still a lot of other files and Sandcastle demos that reference `preRender` and `postRender`, do those need to use the update listeners instead too?

ggetz replied:
Usage of postRender didn't need to be changed in the examples, but I modified the `preRender` events where necessary.

lilleyse commented on Source/Scene/Scene.js:
```diff
@@ -1024,6 +1027,45 @@ define([
+         * @see scene#postUpdate
+         * @see scene#preRender
+         * @see scene#postRender
+         * @see scene#render
```
Capitalize `Scene` for these and below.

lilleyse commented on Source/Scene/Scene.js:
```diff
@@ -1024,6 +1027,45 @@ define([
+         * @see scene#postUpdate
+         * @see scene#preRender
+         * @see scene#postRender
+         * @see scene#render
```
Remove the link to `scene#render`, it is a private function.

lilleyse replied:
Same below.

lilleyse commented on Source/Scene/Scene.js:
```diff
@@ -2787,7 +2840,7 @@ define([
     }
 
-    function callAfterRenderFunctions(frameState) {
+    function callAfterRenderCycleFunctions(frameState) {
```
Why the name change?

ggetz replied:
I wanted to make it clear that it was called every call to `scene.render`, whether a new frame is rendered or not, but it's not necessary in retrospect.

lilleyse commented on Source/Scene/Scene.js:
```diff
@@ -2922,6 +2968,18 @@ define([
+        try {
+            functionToExecute(scene, time);
+        } catch (error) {
+            scene._renderError.raiseEvent(scene, error);
```
Both `update` and `render` use `renderError`. Do we need an `updateError`?

ggetz replied:
The only place it's used in Cesium is that the viewer listens to it, and stops the render loop and displays it, which we'd still want. In case any app is also listening for this event, errors occurring during `update` would no longer be caught, which might be an issue.

Both are occurring during the `scene.render` function, so I don't think the naming is an issue, but if we wanted we could rename and deprecate `scene.renderError`

lilleyse replied:
Probably okay to leave as is for now. Maybe we can reevaluate for the upcoming PR.

lilleyse commented on Specs/Scene/SceneSpec.js:
```diff
@@ -983,7 +983,47 @@ defineSuite([
+        s.destroyForSpecs();
+    });
+
+    it('always raises preUpdate event after updating', function() {
```
This is a duplicate of the other test. Should this be postUpdate?

lilleyse commented:
The recent changes look good.

Would anyone else like to review? The PR is being merged into `cpu-usage` so there will be an opportunity to review everything once all the incremental PRs are done.

lilleyse commented:
Merging into `cpu-usage`.
